libgalois/src/PropertyGraph.cpp

```diff
@@ -551,3 +551,124 @@ katana::SortNodesByDegree(katana::PropertyGraph* pg) {
 
   return katana::ResultSuccess();
 }
+
+katana::Result<std::unique_ptr<katana::PropertyGraph>>
+katana::CreateSymmetricGraph(katana::PropertyGraph* pg) {
+  const GraphTopology& topology = pg->topology();
+  auto out_dests_view_result =
+      ConstructPropertyView<UInt32Property>(topology.out_dests.get());
+  if (!out_dests_view_result) {
+    return out_dests_view_result.error();
+  }
+  //Original property graph destination vector
+  auto out_dests_view = std::move(out_dests_view_result.value());
+
+  // New symmetric graph topology
+  katana::LargeArray<uint64_t>* out_indices_symmetric =
+      new katana::LargeArray<uint64_t>();
+  katana::LargeArray<uint32_t>* out_dests_symmetric =
+      new katana::LargeArray<uint32_t>();
+  katana::LargeArray<uint64_t> out_indices_symmetric_temp;
+
+  out_indices_symmetric->allocateInterleaved(topology.num_nodes());
+  out_indices_symmetric_temp.allocateInterleaved(topology.num_nodes());
+
+  auto edge_range = topology.edge_range(0);
+  (*out_indices_symmetric)[0] = edge_range.first;
+  katana::do_all(
+      katana::iterate(uint64_t{1}, topology.num_nodes()), [&](uint64_t n) {
+        auto edge_range = topology.edge_range(n);
+        (*out_indices_symmetric)[n] = (edge_range.second - edge_range.first);
+      });
+
+  katana::do_all(
+      katana::iterate(uint64_t{0}, topology.num_nodes()),
+      [&](uint64_t n) {
+        auto edge_range = topology.edge_range(n);
+        // update the out_indices for the symmetric topology
+        for (auto e = edge_range.first; e != edge_range.second; ++e) {
+          auto dest = out_dests_view[e];
+          //Ignore self edges
+          if (n != dest) {
+            __sync_fetch_and_add(&((*out_indices_symmetric)[dest]), 1);
```

Comment:
What if this edge already exists? Please document that in the function declaration.